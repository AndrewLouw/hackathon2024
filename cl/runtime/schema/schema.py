# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import importlib
import inspect
from cl.runtime.records.class_info import ClassInfo
from cl.runtime.schema.type_decl import TypeDecl
from cl.runtime.schema.type_decl import pascalize
from cl.runtime.schema.type_decl_key import TypeDeclKey
from cl.runtime.settings.context_settings import ContextSettings
from collections import Counter
from collections import defaultdict
from enum import Enum
from memoization import cached
from pkgutil import walk_packages
from types import ModuleType
from typing import Dict
from typing import Iterable
from typing import List
from typing import Type
from typing_extensions import Self


def is_key_or_record(data_type):
    """Return true if the type is a record based on the presence of 'get_key_type' method."""
    return (
        inspect.isclass(data_type)
        and hasattr(data_type, "get_key_type")
        and not inspect.isabstract(data_type)
        and not data_type.__name__.endswith("Mixin")
    )


class Schema:
    """
    Provide declarations for the specified type and all dependencies.
    """

    _type_dict_by_short_name: Dict[str, Type] = None

    @classmethod
    @cached
    def get_types(cls) -> Iterable[Type]:
        """
        Get all record and field types found in the list of packages specified in settings.

        Notes:
            The result is returned in the alphabetical order of module.ClassName.
        """
        return cls.get_type_dict().values()

    @classmethod
    def add_types_to_dict_by_short_name(cls, types: Iterable[Type]) -> None:
        """Add types to the dictionary by short name (class name with optional package alias)."""

        types_dict = {x.__name__: x for x in types}
        cls.get_type_dict().update(types_dict)

    @classmethod
    def get_type_by_short_name(cls, short_name: str) -> Type:
        """Get type from short name (class name with optional package alias)."""

        # Get dictionary of types indexed by alias
        type_dict_by_short_name = cls.get_type_dict()

        # TODO: Update to support short name with namespace prefix
        record_type = type_dict_by_short_name.get(short_name, None)
        if record_type is None:
            raise RuntimeError(
                f"Record class with short name {short_name} is not found "
                f"in the list of packages specified in settings."
            )
        return record_type

    @classmethod
    def get_type_dict(cls) -> Dict[str, Type]:
        """
        Get dictionary of types indexed by short name (class name with optional package alias).

        Notes:
            The result is returned in the alphabetical order of module.ClassName.
        """

        if cls._type_dict_by_short_name is None:
            # Load packages from Dynaconf
            context_settings = ContextSettings.instance()
            packages = context_settings.packages

            # Get modules for the specified packages
            modules = cls._get_modules(packages)

            # Get record types by iterating over modules
            record_types = set(
                record_type for module in modules for name, record_type in inspect.getmembers(module, is_key_or_record)
            )

            # Ensure names are unique
            # TODO: Support namespace aliases to resolve conflicts
            record_names = [record_type.__name__ for record_type in record_types]
            record_paths = [f"{record_type.__module__}.{record_type.__name__}" for record_type in record_types]

            # Check that there are no repeated names, report errors if there are
            if len(set(record_names)) != len(record_names):
                # Count the occurrences of each name in the list
                record_name_counts = Counter(record_names)

                # Find names that are repeated more than once
                repeated_names = [record_name for record_name, count in record_name_counts.items() if count > 1]

                # Report repeated names
                package_names_str = ", ".join(packages)
                repeated_names_str = ", ".join(repeated_names)
                raise RuntimeError(
                    f"The following class names in the list of packages {package_names_str} "
                    f"are repeated more than once: {repeated_names_str}"
                )

            # Create dictionary
            result = dict(zip(record_names, record_types))

            # Sort alphabetically by module_shortname.ClassName
            # TODO: Support module_shortname
            cls._type_dict_by_short_name = {key: result[key] for key in sorted(result)}

        return cls._type_dict_by_short_name

    @classmethod
    def for_key(cls, key: TypeDeclKey) -> Self:
        """Create or return cached object for the specified type declaration key."""
        class_path = f"{key.module.module_name}.{key.name}"
        return cls.for_class_path(class_path)

    @classmethod
    def for_class_path(cls, class_path: str) -> Dict[str, Dict]:
        """Create or return cached object for the specified class path in module.ClassName format."""

        record_type = ClassInfo.get_class_type(class_path)
        return cls.for_type(record_type)

    @classmethod
    @cached
    def for_type(cls, record_type: Type) -> Dict[str, Dict]:
        """
        Declarations for the specified type and all dependencies, returned as a dictionary.

        Args:
            record_type: Type of the record for which the schema is created.
        """
        dependencies = set()

        # Get or create type declaration the argument class
        type_decl_obj = TypeDecl.for_type(record_type, dependencies=dependencies)

        # Sort the list of dependencies
        # TODO: Use short name with alias
        dependencies = sorted(dependencies, key=lambda x: x.__name__)

        # Add to the dict by type name
        cls.add_types_to_dict_by_short_name(dependencies)

        # TODO: Restore after Enum decl generation is supported
        dependencies = [dependency_type for dependency_type in dependencies if not issubclass(dependency_type, Enum)]

        # Requested type is always first
        type_decl_list = [type_decl_obj] + [TypeDecl.for_type(dependency_type) for dependency_type in dependencies]

        # TODO: Move pascalize to a helper class
        result = {
            pascalize(f"{type_decl.module.module_name}.{type_decl.name}"): type_decl.to_type_decl_dict()
            for type_decl in type_decl_list
        }
        return result

    @classmethod
    @cached
    def _get_modules(cls, packages: List[str]) -> List[ModuleType]:
        """
        Get a list of ModuleType objects for submodules at all levels of the specified packages or root modules
        in the alphabetical order of dot-delimited module name.

        Args:
            packages: List of packages or root module strings in dot-delimited format, for example ['cl.runtime']
        """
        result = []
        for package in packages:
            # Import root module of the package
            root_module = importlib.import_module(package)
            result.append(root_module)  # Add the root module itself
            # Get module info for all submodules, note the trailing period added as per walk_packages documentation
            for module_info in walk_packages(root_module.__path__, root_module.__name__ + "."):
                module_name = module_info.name
                # Import the submodule using its full name
                submodule = importlib.import_module(module_name)
                result.append(submodule)

        # Sort the result by module path
        result = sorted(result, key=lambda module: module.__name__)
        return result

    @classmethod
    @cached
    def get_types_in_hierarchy(cls, record_type: Type) -> List[Type]:
        """
        Find all record types in hierarchy for given record_type.
        Include all base and child classes ordered by hierarchy.
        """

        get_key_type = getattr(record_type, "get_key_type", None)

        if get_key_type is None:
            raise RuntimeError(f"Type {record_type} is not record type.")

        # get key type
        # TODO (Roman): review get_key_type method of KeyProtocol. maybe it can be staticmethod or classmethod?
        key_type = get_key_type(None)

        # container to collect types
        types_ = defaultdict(list)

        for type_ in cls.get_types():
            # TODO (Roman): should not skip if input type is not key
            if type_ == record_type:
                # skip start type
                continue

            type_get_key_type = getattr(type_, "get_key_type", None)
            if type_get_key_type is None:
                # skip non-record types
                continue

            try:
                type_key_type = type_get_key_type(None)
            except AttributeError:
                # skip types whose key type unavailable from type
                continue

            if key_type != type_key_type:
                # skip types of another key type
                continue

            # try to resolve hierarchy
            try:
                i = type_.__mro__.index(key_type)
            except ValueError:
                # if key_type is not in mro default i = 1
                i = 1

            types_[i].append(type_)

        # final list of types in hierarchy
        result: List[Type] = []

        # order by place in hierarchy. more derived in the end.
        for k, v in sorted(types_.items(), key=lambda item: item[0]):
            result.extend(v)

        return result
