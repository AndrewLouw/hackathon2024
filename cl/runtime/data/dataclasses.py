# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import dataclasses
from typing import Any, Optional
from typing_extensions import dataclass_transform


def data_field(
    *,
    default: Optional[Any] = None,
    factory: Optional[Any] = None,
    optional: bool = False,
    optional_fields: bool = True,
    subtype: Optional[str] = None,
    name: Optional[str] = None,
    label: Optional[str] = None,
    formatter: Optional[str] = None,
    category: Optional[str] = None,
    secure: bool = False,
    filterable: bool = False,
) -> Any:
    """Field in dataclass with additional parameters to define runtime-specific metadata.

    Args:
        default: Default value (None if not specified)
        factory: Factory to generate a new instance for default value (for container types)
        optional: If not specified, the field must be set before saving the record.
        optional_fields: Whether the elements of list are optional.
        subtype: Subtype of the field type. Permitted values are `long` for int fields,
        `date` and `datetime` for string fields. TODO(attrs) - Review the list, determine which strings to convert
        name: Name of the field in DB, use when standard name conversion is inapplicable due to Python constraints.
        label: Readable name when not obtained by the standard conversion rules.
        formatter: Standard formatter name (without curly brackets) or raw Python format string (in curly brackets)
        category: A group of fields displayed together in the UI. Has no effect outside UI.
        secure: Marks the field as secure TODO(attrs) - Explain further in docs
        filterable: Marks the field as filterable TODO(attrs) - Explain further in docs
    """
    metadata = {
        'data_field': True,
        'optional': optional,
        'optional_fields': optional_fields,
        'type': subtype,
        'name': name,
        'label': label,
        'format': formatter,  # TODO(attrs) - switch to formatter in remaining code as format causes Python warning
        'category': category,
        'secure': secure,
        'filterable': filterable,
    }
    if factory is None:
        return dataclasses.field(default=default, metadata=metadata)
    elif default is None:
        return dataclasses.field(default_factory=factory, metadata=metadata)
    else:
        raise RuntimeError(f"Fields default={default} and factory={factory} in data_class are mutually exclusive.")


@dataclass_transform()
def data_class_impl(cls, *, init: bool = True, label: str = None):
    """Performs the actual wrapping irrespective of call syntax with or without parentheses."""

    cls = dataclasses.dataclass(cls)

    # TODO: Remove base fields
    # fields = {f.name: f for f in dataclasses.fields(cls) if not f.inherited}

    # TODO: Apply init parameter
    # TODO: Apply label parameter

    def to_dict(self):
        return dataclasses.asdict(self)

    cls.to_dict = to_dict

    if not hasattr(cls, "init"):
        def init(self):
            pass

        cls.init = init

    return cls


@dataclass_transform()
def data_class(cls=None, *, init: bool = True, label: str = None):
    """Runtime decorator dataclasses."""

    # The value of cls type depends on whether parentheses follow the decorator.
    # It is the class when used as @data_class but None for @data_class().
    if cls is None:
        return data_class_impl
    else:
        return data_class_impl(cls, init=init, label=label)
