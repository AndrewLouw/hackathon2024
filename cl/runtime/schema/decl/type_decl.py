# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import List, Optional
from cl.runtime.decorators.attrs_record_decorator import attrs_record
from cl.runtime.schema.decl.handler_declare_block_decl import HandlerDeclareBlockDecl
from cl.runtime.schema.decl.handler_implement_block_decl import HandlerImplementBlockDecl
from cl.runtime.schema.decl.interface_decl_key import InterfaceDeclKey
from cl.runtime.schema.decl.type_argument_decl import TypeArgumentDecl
from cl.runtime.schema.decl.type_decl_key import TypeDeclKey
from cl.runtime.schema.decl.type_element_decl import TypeElementDecl
from cl.runtime.schema.decl.type_index_decl import TypeIndexDecl
from cl.runtime.schema.decl.type_kind import TypeKind
from cl.runtime.schema.decl.type_param_decl import TypeParamDecl
from cl.runtime.decorators.data_field_decorator import data_field


@attrs_record
class TypeDecl(TypeDeclKey):
    """
    Defines type declaration. A tag of entity type XML representation corresponds to each element of the type. The
    names of type elements and corresponding tags coincide.
    """

    label: str = data_field()
    """Type label."""

    comment: Optional[str] = data_field()
    """Type comment. Contains additional information."""

    category: Optional[str] = data_field()
    """Category."""

    shortcut: Optional[str] = data_field()
    """Shortcut."""

    type_params: Optional[List[TypeParamDecl]] = data_field()
    """Type Params"""

    aliases: Optional[List[str]] = data_field()
    """Type aliases."""

    kind: Optional[TypeKind] = data_field()
    """Type kind."""

    inherit: Optional[TypeDeclKey] = data_field()
    """Parent type reference."""

    inherit_type_arguments: Optional[List[TypeArgumentDecl]] = data_field()
    """Inherit Type Argument."""

    interfaces: Optional[List[InterfaceDeclKey]] = data_field()
    """Parent interfaces"""

    declare: Optional[HandlerDeclareBlockDecl] = data_field()
    """Handler declaration block."""

    implement: Optional[HandlerImplementBlockDecl] = data_field()
    """Handler implementation block."""

    elements: Optional[List[TypeElementDecl]] = data_field()
    """Element declaration block."""

    keys: Optional[List[str]] = data_field()
    """Array of key element names."""

    indexes: Optional[List[TypeIndexDecl]] = data_field()
    """Defines indexes for the type."""

    immutable: Optional[bool] = data_field()
    """Immutable flag."""

    ui_response: Optional[bool] = data_field()
    """Flag indicating UiResponse."""

    seed: Optional[int] = data_field()
    """Seed."""

    version: Optional[str] = data_field()
    """Type Version"""

    system: Optional[bool] = data_field()
    """System."""

    enable_cache: Optional[bool] = data_field()
    """Enable cache flag."""

    partial: Optional[bool] = data_field()
    """It is possible to split the code over two or more source files."""

    abstract: Optional[bool] = data_field()
    """
    Abstract flag: is a restricted type that cannot be used to create objects (to access it, it must be inherited from
    another type).
    """

    interactive_edit: Optional[bool] = data_field()
    """Interactive Edit"""

    permanent: Optional[bool] = data_field()
    """Save records always permanently."""

    pinned: Optional[bool] = data_field()
    """Store records always in root dataset."""
