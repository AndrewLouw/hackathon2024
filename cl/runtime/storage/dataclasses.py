# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Any, Optional
from typing_extensions import dataclass_transform

from cl.runtime.storage.attrs import attrs_field, attrs_class, attrs_class_impl


def data_field(
    *,
    default: Optional[Any] = None,
    factory: Optional[Any] = None,
    optional: bool = False,
    optional_fields: bool = True,
    subtype: Optional[str] = None,
    name: Optional[str] = None,
    label: Optional[str] = None,
    formatter: Optional[str] = None,
    category: Optional[str] = None,
    secure: bool = False,
    filterable: bool = False,
) -> Any:
    """Field in dataclass with additional parameters to define runtime-specific metadata.

    Args:
        default: Default value (None if not specified)
        factory: Factory to generate a new instance for default value (for container types)
        optional: If not specified, the field must be set before saving the record.
        optional_fields: Whether the elements of list are optional.
        subtype: Subtype of the field type. Permitted values are `long` for int fields,
        `date` and `datetime` for string fields. TODO(attrs) - Review the list, determine which strings to convert
        name: Name of the field in DB, use when standard name conversion is inapplicable due to Python constraints.
        label: Readable name when not obtained by the standard conversion rules.
        formatter: Standard formatter name (without curly brackets) or raw Python format string (in curly brackets)
        category: A group of fields displayed together in the UI. Has no effect outside UI.
        secure: Marks the field as secure TODO(attrs) - Explain further in docs
        filterable: Marks the field as filterable TODO(attrs) - Explain further in docs
    """
    return attrs_field(
        default=default,
        factory=factory,
        optional=optional,
        optional_fields=optional_fields,
        subtype=subtype,
        name=name,
        label=label,
        formatter=formatter,
        category=category,
        secure=secure,
        filterable=filterable
        )


@dataclass_transform()
def data_class_impl(cls, *, init: bool = True, label: str = None):
    """Performs the actual wrapping irrespective of call syntax with or without parentheses."""
    return attrs_class_impl(cls=cls, init=init, label=label)


@dataclass_transform()
def data_class(cls=None, *, init: bool = True, label: str = None):
    """Runtime decorator dataclasses."""
    return attrs_class(cls=cls, init=init, label=label)
