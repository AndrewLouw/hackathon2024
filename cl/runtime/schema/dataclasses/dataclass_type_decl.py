# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import dataclasses

from memoization import cached
from typing_extensions import Self
from cl.runtime.schema.dataclasses.dataclass_field_decl import DataclassFieldDecl
from cl.runtime.schema.type_decl import TypeDecl, for_type_key_maker
from cl.runtime.schema.element_decl import ElementDecl
from dataclasses import dataclass
from typing import get_type_hints, Type


@dataclass(slots=True, kw_only=True)
class DataclassTypeDecl(TypeDecl):
    """Type declaration for a dataclass."""

    @classmethod
    @cached(custom_key_maker=for_type_key_maker)
    def for_type(cls, record_type: Type, *, skip_fields: bool = False) -> Self:
        """
        Create or return cached object for the specified record type.

        Args:
            record_type: Type of the record for which the declaration is created
            skip_fields: Use this flag to skip fields generation when the method is invoked from a derived class
        """

        if not dataclasses.is_dataclass(record_type):
            raise RuntimeError(f"DataclassTypeDecl used for {record_type.__name__} which is not a dataclass.")

        # Populate using TypeDecl base
        result = TypeDecl.for_type(record_type, skip_fields=True)

        # Use this flag to skip fields generation when the method is invoked from a derived class
        if not skip_fields:

            # Information about dataclass fields including the metadata (does not resolve ForwardRefs)
            fields = dataclasses.fields(cls)

            # Get type hints to resolve ForwardRefs
            type_hints = get_type_hints(cls)

            # Add elements
            result.elements = []
            for field in fields:

                # Get type from type hints because they resolve forward references
                field_type = type_hints[field.name]

                # Get the rest of the data from the field itself
                field_decl = DataclassFieldDecl.create(field, field_type)

                # Convert to element and add
                element_decl = ElementDecl.create(field_decl)
                result.elements.append(element_decl)

        return result
