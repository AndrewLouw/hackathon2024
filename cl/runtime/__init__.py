from cl.runtime.decorators.handler_decorator import handler
from cl.runtime.decorators.viewer_decorator import viewer
from cl.runtime.classes.data_mixin import DataMixin
from cl.runtime.classes.key_mixin import KeyMixin
from cl.runtime.classes.record_mixin import RecordMixin
from cl.runtime.classes.attrs import data_class, data_field
from cl.runtime.classes.record_util import RecordUtil
from cl.runtime.storage.data_source import DataSource
from cl.runtime.storage.context import Context
from cl.runtime.storage.in_process.in_process_data_source import InProcessDataSource
from cl.runtime.view.view import View
from cl.runtime.view.record_view import RecordView
from cl.runtime.view.record_list_view import RecordListView
from cl.runtime.schema.type.package_decl_key import PackageDeclKey
from cl.runtime.schema.type.package_decl import PackageDecl
from cl.runtime.schema.type.type_decl_key import TypeDeclKey
from cl.runtime.schema.type.type_decl import TypeDecl
from cl.runtime.schema.type.field_decl import FieldDecl
from cl.runtime.schema.type.primitive_decl import PrimitiveDecl
from cl.runtime.schema.type.enum_decl import EnumDecl
from cl.runtime.schema.type.data_decl import DataDecl
from cl.runtime.storage.class_info import ClassInfo
