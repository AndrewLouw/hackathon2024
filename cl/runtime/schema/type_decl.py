# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import ast
import inspect
from itertools import tee

from cl.runtime import KeyUtil
from cl.runtime.records.dataclasses.dataclass_mixin import DataclassMixin
from cl.runtime.records.dataclasses.dataclass_mixin import datafield
from cl.runtime.schema.element_decl import ElementDecl
from cl.runtime.schema.field_decl import FieldDecl
from cl.runtime.schema.handler_declare_block_decl import HandlerDeclareBlockDecl
from cl.runtime.schema.module_decl_key import ModuleDeclKey, ModuleDeclTable
from cl.runtime.schema.type_decl_key import TypeDeclKey, TypeDeclTable
from cl.runtime.schema.type_index_decl import TypeIndexDecl
from cl.runtime.schema.type_kind import TypeKind
from dataclasses import dataclass, asdict
from enum import Enum
from inflection import titleize, camelize
from memoization import cached
from typing import Dict, Literal, Any, Set
from typing import List
from typing import Type
from typing import get_type_hints
from typing_extensions import Self

DisplayKindLiteral = Literal["Basic", "Singleton", "Dashboard"]


def pascalize(s: str) -> str:
    """Split into dot-delimited tokens, pascalize each token, then concatenate."""
    tokens = s.split(".")
    tokens = [camelize(token, uppercase_first_letter=True) for token in tokens]
    result = ".".join(tokens)
    return result


def to_type_decl_dict(node: Dict[str, Any] | List[Dict[str, Any]] | str) -> Dict[str, Any] | List[Dict[str, Any]] | str:
    """Recursively apply type declaration dictionary conventions to the input."""

    if isinstance(node, dict):
        # For type declarations only, skip nodes that have the value of None or False
        # Remove suffix _ from field names if present
        # pascalized_values = {k: (pascalize(v) if k in ['module_name', 'name'] else v) for k, v in node.items()}
        result = {pascalize(k.removesuffix("_")): to_type_decl_dict(v) for k, v in node.items() if v not in [None, False]}
        return result
    elif isinstance(node, list):
        # For type declarations only, skip nodes that have the value of None or False
        return [to_type_decl_dict(v) for v in node if v not in [None, False]]
    elif isinstance(node, tuple):
        # The first element of key node tuple is type, the remaining elements are primary key fields
        # Remove suffix _ from field names if present
        key_field_names = node[0].get_key_fields()
        key_field_values = [to_type_decl_dict(v) for v in node[1:]]
        return {pascalize(k.removesuffix("_")): v for k, v in zip(key_field_names, key_field_values)}
    elif isinstance(node, str):
        return pascalize(node)
    else:
        return node


def for_type_key_maker(cls, record_type: Type, *, dependencies: Set[Type] | None = None, skip_fields: bool = False) -> str:
    """Custom key marker for 'for_type' class method."""
    # TODO: Replace by lambda if skip_fields parameter is removed
    return f"{record_type.__module__}.{record_type.__name__}.{dependencies.__hash__}{skip_fields}"


@dataclass(slots=True, kw_only=True)
class TypeDecl(DataclassMixin):
    """Provides information about a class, its fields, and its methods."""

    module: ModuleDeclKey = datafield()  # TODO: Merge with name to always use full name
    """Module reference."""

    name: str = datafield()
    """Type name is unique when combined with module."""

    label: str | None = datafield()
    """Type label."""

    comment: str | None = datafield()
    """Type comment. Contains additional information."""

    kind: TypeKind | None = datafield()
    """Type kind."""

    display_kind: DisplayKindLiteral = datafield()  # TODO: Make optional, treat None as Basic
    """Display kind."""

    inherit: TypeDeclKey | None = datafield()
    """Parent type reference."""

    declare: HandlerDeclareBlockDecl | None = datafield()  # TODO: Flatten or use block for abstract flag
    """Handler declaration block."""

    elements: List[ElementDecl] | None = datafield()  # TODO: Consider renaming to fields
    """Element declaration block."""

    keys: List[str] | None = datafield()
    """Array of key element names (specify in base class only)."""

    indexes: List[TypeIndexDecl] | None = datafield()
    """Defines indexes for the type."""

    immutable: bool | None = datafield()
    """Immutable flag."""

    permanent: bool | None = datafield()
    """When the record is saved, also save it permanently."""

    def get_key(self) -> TypeDeclKey:
        return TypeDeclTable, self.module, self.name

    def to_type_decl_dict(self) -> Dict[str, Any]:
        """Convert to dictionary using type declaration conventions."""

        # Convert to standard dictionary format
        standard_dict = asdict(self)

        # Apply type declaration dictionary conventions
        result = to_type_decl_dict(standard_dict)
        return result

    @classmethod
    def for_key(cls, key: TypeDeclKey) -> Self:
        """Create or return cached object for the specified type declaration key."""
        class_path = f"{key[1][1]}.{key[2]}"  # TODO: Use parse_key method
        return cls.for_class_path(class_path)

    @classmethod
    def for_class_path(cls, class_path: str) -> Self:
        """Create or return cached object for the specified class path in module.ClassName format."""
        raise NotImplementedError()

    @classmethod
    @cached(custom_key_maker=for_type_key_maker)
    def for_type(cls, record_type: Type, *, dependencies: Set[Type] | None = None, skip_fields: bool = False) -> Self:
        """
        Create or return cached object for the specified record type.

        Args:
            record_type: Type of the record for which the declaration is created
            dependencies: Set of types used in field or methods of the specified type, populated only if not None
            skip_fields: Use this flag to skip fields generation when the method is invoked from a derived class
        """

        if issubclass(record_type, Enum):
            raise RuntimeError(f"Cannot create TypeDecl for class {record_type.__name__} because it is an enum.")
        if issubclass(record_type, tuple):
            raise RuntimeError(f"Cannot create TypeDecl for class {record_type.__name__} because it is a tuple.")

        # Create instance of the final type
        result = cls()

        result.module = ModuleDeclTable.create_key(module_name=record_type.__module__)
        result.name = record_type.__name__
        result.label = titleize(result.name)  # TODO: Add override from settings
        result.comment = record_type.__doc__

        # Set type kind by detecting the presence of 'get_key' method to indicate a record vs. an element
        is_record = hasattr(record_type, "get_key")
        is_abstract = hasattr(record_type, "__abstractmethods__") and bool(record_type.__abstractmethods__)
        if is_record:
            result.kind = TypeKind.Abstract if is_abstract else None
        else:
            result.kind = TypeKind.AbstractElement if is_abstract else TypeKind.Element

        # Set display kind
        result.display_kind = "Basic"  # TODO: Remove Basic after display_kind is made optional

        # Set parent class as the first class in MRO that is not self and does not have Mixin suffix
        for parent_type in record_type.__mro__:
            if parent_type is not record_type and not parent_type.__name__.endswith("Mixin"):
                # TODO: Add to dependencies
                parent_type_module = ModuleDeclTable.create_key(module_name=parent_type.__module__)
                parent_type_name = parent_type.__name__
                # TODO: result.inherit = TypeDeclTable.create_key(module=parent_type_module, name=parent_type_name)

        # Get key fields by parsing the source of 'get_key' method
        result.keys = KeyUtil.get_key_fields(record_type)

        # Use this flag to skip fields generation when the method is invoked from a derived class
        if not skip_fields:
            # Get type hints to resolve ForwardRefs
            type_hints = get_type_hints(record_type)

            # Dictionary of member comments (docstrings), currently requires source parsing due Python limitations
            member_comments = cls.get_member_comments(record_type)

            # Add an element for each type hint
            result.elements = []
            for field_name, field_type in type_hints.items():

                # Field comment (docstring)
                field_comment = member_comments.get(field_name, None)

                # Get the rest of the data from the field itself
                field_decl = FieldDecl.create(record_type, field_name, field_type, field_comment, dependencies=dependencies)

                # Convert to element and add
                element_decl = ElementDecl.create(field_decl)
                result.elements.append(element_decl)

        return result

    @classmethod
    @cached
    def get_member_comments(cls, record_type: type) -> Dict[str, str]:
        """Extract class member comments."""

        comments = dict()
        ast_tree = ast.parse(inspect.getsource(record_type))

        for i, j in cls.by_pair(ast.iter_child_nodes(ast_tree.body[0])):
            if isinstance(i, ast.AnnAssign):
                target_node = i.target
            elif isinstance(i, ast.Assign):
                target_node = i.targets[0]
            else:
                continue

            if not isinstance(target_node, ast.Name):
                continue

            name: str = target_node.id
            # TODO: ast.Str is replaced by ast.Constant in Python 3.8, update
            if isinstance(j, ast.Expr) and isinstance(j.value, ast.Str):
                comments[name] = inspect.cleandoc(j.value.s)

        return comments

    @classmethod
    def by_pair(cls, iterable):
        """s -> (s0,s1), (s1,s2), (s2, s3), ..."""
        a, b = tee(iterable)
        next(b, None)
        return zip(a, b)
