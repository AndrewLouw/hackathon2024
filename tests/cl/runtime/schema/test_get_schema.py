# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import dataclasses
import json
import os
import pytest
from cl.runtime.schema.dataclasses.dataclass_type_decl import DataclassTypeDecl
from cl.runtime.schema.type_decl import TypeDecl
from stubs.cl.runtime import StubDataclassNestedFields, StubCustomBase, StubDataclassListFields, \
    StubDataclassPrimitiveFields
from stubs.cl.runtime import StubDataclassRecord
from stubs.cl.runtime.records.dataclasses.stub_dataclass_optional_fields import StubDataclassOptionalFields


def test_method():
    """Test coroutine for /schema/typeV2 route."""

    sample_types = [
        StubDataclassRecord,
        StubDataclassPrimitiveFields,
        StubDataclassListFields,
        StubDataclassNestedFields,
        StubDataclassOptionalFields
    ]

    for sample_type in sample_types:
        class_module = sample_type.__module__.rsplit(".", maxsplit=1)[1]
        received_result_file_path = os.path.abspath(__file__).replace(".py", f".{class_module}.received.json")
        expected_result_file_path = os.path.abspath(__file__).replace(".py", f".{class_module}.expected.json")

        result_obj = DataclassTypeDecl.for_type(sample_type)
        result_dict = result_obj.to_type_decl_dict()

        # Save the dictionary to a file
        with open(received_result_file_path, 'w') as received_result_file:
            json.dump(result_dict, received_result_file, indent=4)

        # Load expected result from a file
        with open(expected_result_file_path, "r", encoding="utf-8") as expected_result_file:
            expected_result_dict = json.load(expected_result_file)

        # Compare
        assert result_dict == expected_result_dict
        pass


if __name__ == "__main__":
    pytest.main([__file__])
