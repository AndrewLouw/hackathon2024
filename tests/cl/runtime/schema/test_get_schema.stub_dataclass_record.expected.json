{
    "module": {
        "module_name": "stubs.cl.runtime.records.dataclasses.stub_dataclass_record"
    },
    "name": "StubDataclassRecord",
    "label": "Stub Dataclass Record",
    "comment": "Stub record base class.",
    "display_kind": "Basic",
    "elements": [
        {
            "value": {
                "type": "str"
            },
            "name": "id",
            "comment": "Unique identifier.",
            "optional": false
        }
    ],
    "keys": [
        "id"
    ]
}